base/special/rem_pio2.jl

```diff
@@ -0,0 +1,284 @@
+
+    # 2. Let α = 1/2π, then:
+    #
+    #    α*x mod 1 ≡ [(α*2^k mod 1)*X] mod 1
+    #
+    # so we can ignore the first k bits of α. Extract the next 3 64-bit parts of α.
+    #
+    # i.e. equivalent to
+    #     setprecision(BigFloat,4096)
+    #     α  = 1/(2*big(pi))
+    #     A  = mod(ldexp(α,k), 1)
+    #     z1 = ldexp(A,64)
+    #     a1 = trunc(UInt64, z1)
+    #     z2 = ldexp(z1-a1, 64)
+    #     a2 = trunc(UInt64, z2)
+    #     z3 = ldexp(z2-a2, 64)
+    #     a3 = trunc(UInt64, z3)
+
+    # idx, shift = divrem(k, 64), but divrem is slower
+    idx = k >> 6
+    shift = k - (idx << 6)
+    if shift == 0
+        a1 = INV2PI[idx+1]
+        a2 = INV2PI[idx+2]
+        a3 = INV2PI[idx+3]
+    else
+        # use shifts to extract the relevant 64 bit window
+        a1 = (idx < 0 ? zero(UInt64) : INV2PI[idx+1] << shift) | (INV2PI[idx+2] >> (64 - shift))
+        a2 = (INV2PI[idx+2] << shift) | (INV2PI[idx+3] >> (64 - shift))
+        a3 = (INV2PI[idx+3] << shift) | (INV2PI[idx+4] >> (64 - shift))
+    end
+
+    # 3. Perform the multiplication:
+    #
+    #      X.  0  0  0
+    #   ×  0. a1 a2 a3
+    #   ==============
+    #      _.  w  w  _
+    #
+    # (i.e. ignoring integer and lowest bit parts of result)
+
+    w1 = UInt128(X*a1) << 64 # overflow becomes integer
+    w2 = widemul(X,a2)
+    w3 = widemul(X,a3) >> 64
+    w = w1 + w2 + w3         # quotient fraction after division by 2π
+
+    # adjust for sign of x
+    w = flipsign(w,x)
+
+    # 4. convert to quadrant, quotient fraction after division by π/2:
+    q = (((w>>125)%Int +1)>>1) # nearest quadrant
+    f = (w<<2) % Int128 # fraction part of quotient after division by π/2, taking values on [-0.5,0.5)
+
+    # 5. convert quotient fraction to split precision Float64
+    z_hi,z_lo = fromfraction(f)
+
+    # 6. multiply by π/2
+    pio2_hi = 1.5707963407039642
+    pio2_lo = -1.3909067614167116e-8
+
+    y_hi = (z_hi+z_lo)*(pio2_hi+pio2_lo)
+    y_lo = (((z_hi*pio2_hi - y_hi) + z_hi*pio2_lo) + z_lo*pio2_hi) + z_lo*pio2_lo
+    return q, y_hi, y_lo
+end
+
+
+"""
+    highword(x)
+
+returns the high word of x as a UInt32.
+"""
+@inline highword(x::UInt64) = unsafe_trunc(UInt32,x >> 32)
+@inline highword(x::Float64) = unsafe_trunc(UInt32,highword(reinterpret(UInt64, x)))
+
+"""
+    rem_pio2(x::Float64)
+
+returns the remainder of x modulo π/2 as a TwicePrecision number, along with a k
+such that k mod 3 == K mod 3 where K*π/2 = x -rem.
+"""
+
+function rem_pio2(x::Float64)
+    xh = highword(x)
+    xhp = xh & 0x7fffffff # positive part of highword
+
+    if xhp <= 0x3fe921fb #|x| ~<= pi/4, no need for reduction, just return input
+        return Int(0), x, 0.0
+    end
+    rem_pio2_kernel(x, xh, xhp)
+end
+
+"""
+    rem_pio2_kernel(x, xh, xhp)
+
+returns the remainder of x modulo π/2 as a TwicePrecision number, along with a k
+such that k mod 3 == K mod 3 where K*π/2 = x -rem.
```

Comment:
maybe some code / math highlighting here? is this supposed to be `x - rem` or something else?